### PR TITLE
JBDS-3951 Installer needs to add the VirtualBox install directory to the PATH

### PR DESCRIPTION
**Pull request by sudhirverma**
Installer needs to add the VirtualBox install directory to the PATH

### CODE REVIEW

**Comment by sudhirverma** [comment]
test this

**Comment by sudhirverma** [comment]
test this

File: test/unit/model/virtualbox-test.js [hunk 1]
```diff
@@ -240,10 +241,17 @@ describe('Virtualbox installer', function() {
       }
     });
 
+    it('should able to check file were extracted to temp', function(done) {
+      sandbox.stub(child_process, 'execFile').yields(null, "Files were extracted to temp", "");
+      item2.setInstallComplete();
+      item2.thenInstall(installer);
+      installer.install(fakeProgress, success, failure);
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: call installer.installAfterRequirements() and you would not need to create jdk installer at all.

File: test/unit/model/virtualbox-test.js [hunk 2]
```diff
@@ -240,10 +241,17 @@ describe('Virtualbox installer', function() {
...
+      done();
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: if you use done callback from mocha, you should be calling it from callback passed to installer.installAfterRequirements().

``` JavaScript
installer.installAfterRequirements(fakeProgress, function (){
 expect().something.has.been.done;
 done();
 }, function(){
  expect.fail();
 });
```
that is needed only if you really expect async code is still going to be running when test method exits.

if you mock all async methods then you don't need use done callback at all, just `expect().test.the.condition;` should do the trick.

File: test/unit/model/virtualbox-test.js [hunk 3]
```diff
@@ -240,10 +241,17 @@ describe('Virtualbox installer', function() {
       }
     });
 
+    it('should able to check file were extracted to temp', function(done) {
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: Did you mean to name it as 'should run virtualbox installer executable to extract msi installer into temp folder'?

File: browser/model/virtualbox.js [hunk 4]
```diff
@@ -131,6 +132,9 @@ class VirtualBoxInstallWindows extends VirtualBoxInstall {
           return this.configure(installer);
         })
         .then((result) => {
+          if(result && this.targetDir) {
```
Comment (dgolovin) [inline browser/model/virtualbox.js]: just replace to 
``` JavaScript
Platform.addToUserPath(this.getOption('install').location);
```

File: browser/model/virtualbox.js [hunk 5]
```diff
@@ -173,10 +177,10 @@ class VirtualBoxInstallWindows extends VirtualBoxInstall {
       // msiexec logs are in UCS-2
       Util.findText(path.join(this.installerDataSvc.installDir(), 'vbox.log'), 'CopyDir: DestDir=', 'ucs2').then((result)=>{
         let regexTargetDir = /CopyDir: DestDir=(.*)\,.*/;
```
Comment (dgolovin) [inline browser/model/virtualbox.js]: these changes don't needed here

File: browser/model/virtualbox.js [hunk 6]
```diff
@@ -16,6 +16,7 @@ class VirtualBoxInstall extends InstallableItem {
     this.minimumVersion = version;
     this.maximumVersion = '5.2.0';
     this.revision = revision;
+    this.targetDir = "";
```
Comment (dgolovin) [inline browser/model/virtualbox.js]: You don't need new class field, see my comments below. Just use location property from 'install' option as it is explained below.

File: test/unit/model/virtualbox-test.js (hunk 3 above)
Comment (sudhirverma) [inline test/unit/model/virtualbox-test.js]: yes

File: browser/model/virtualbox.js (hunk 4 above)
Comment (sudhirverma) [inline browser/model/virtualbox.js]: Can I use this.getOption('install').location because getOption is not even declare any where.

Comment (dgolovin) [inline browser/model/virtualbox.js]:
``` JavaScript
this.option('install').location
```

**Comment by sudhirverma** [comment]
test this

**Comment by sudhirverma** [comment]
test this

File: browser/services/platform.js [hunk 7]
```diff
@@ -136,7 +136,7 @@ class Platform {
   static addToUserPath_win32(locations) {
     return Platform.getUserPath_win32().then((pathString)=>{
       let pathLocations = pathString.split(';');
-      return Platform.setUserPath_win32([...locations.filter(item=>!pathLocations.includes(item)), pathString].join(';'));
+      return Platform.setUserPath_win32([...[locations].filter(item=>!pathLocations.includes(item)), pathString].join(';'));
```
Comment (dgolovin) [inline browser/services/platform.js]: locations parameter is an array, lets keep it that way. I want to be able to add several locations in single call.

File: browser/model/virtualbox.js [hunk 8]
```diff
@@ -131,6 +131,9 @@ class VirtualBoxInstallWindows extends VirtualBoxInstall {
           return this.configure(installer);
         })
         .then((result) => {
+          if(result) {
```
Comment (dgolovin) [inline browser/model/virtualbox.js]: This check is not required, installation ether succeeded and then callback is executed or failed and catch callback is executed.

File: browser/model/virtualbox.js [hunk 9]
```diff
@@ -131,6 +131,9 @@ class VirtualBoxInstallWindows extends VirtualBoxInstall {
...
+            Platform.addToUserPath(this.option['install'].location);
```
Comment (dgolovin) [inline browser/model/virtualbox.js]: argument must be an array
``` JavaScript
Platform.addToUserPath([this.option['install'].location]);
```

File: test/unit/model/virtualbox-test.js [hunk 10]
```diff
@@ -240,10 +240,23 @@ describe('Virtualbox installer', function() {
       }
     });
 
+    it('should run virtualbox installer executable to extract msi installer into temp folder', function(done) {
+      let stub = sandbox.stub(child_process, 'execFile').yields(null, "Files were extracted to temp", "");
+      try {
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: Try/catch is not required here, please remove it.

File: test/unit/model/virtualbox-test.js [hunk 11]
```diff
@@ -240,10 +240,23 @@ describe('Virtualbox installer', function() {
...
+        installer.install(fakeProgress, success, failure);
+        done();
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: calling done inside main thread is wrong, it should be inside callback, please remove it here and form it second parameter callback.

File: test/unit/model/virtualbox-test.js [hunk 12]
```diff
@@ -240,10 +240,23 @@ describe('Virtualbox installer', function() {
...
+      } catch (error) {
+        expect.fail();
+      }
+      return helper.execFile().then(() => {
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: if you stub everything properly all the work should be done inside main thread an you can just check with expect() here, there is no need to return promise instance from it.

File: test/unit/model/virtualbox-test.js [hunk 13]
```diff
@@ -240,10 +240,23 @@ describe('Virtualbox installer', function() {
...
+        expect(stub).to.be.calledOnce;
+        expect(stub).to.be.true;
```
Comment (dgolovin) [inline test/unit/model/virtualbox-test.js]: this looks wrong, you instead should be checking that stubbed method called with right parameters. In your case you should check that downloaded executable file is called to unpack to temp folder.

**Comment by sudhirverma** [comment]
test this

**Comment by sudhirverma** [comment]
test this

**Comment by dgolovin** [comment]
Included into https://github.com/redhat-developer-tooling/developer-platform-install/pull/544.

**Pull request closed without merging**